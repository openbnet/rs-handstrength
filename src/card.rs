#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}