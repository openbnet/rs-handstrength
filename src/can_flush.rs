use crate::card::{Card, Suit};
use crate::sorting::sort_cards;
use crate::can_libs::*;
use std::collections::HashMap;
// does not care about straight flushes and will return a str flush match as a flush
pub fn can_have_flush(comcards: &Vec<Card>) -> CanHaveCombis {
    if let Some(flush_suit) = can_flush(comcards) {
        let flush_suit_cards: Vec<Card> = comcards.iter()
            .filter(|card| card.suit == flush_suit)
            .cloned()
            .collect();

        let sorted_flush_suit_cards = sort_cards(&flush_suit_cards);

        let unused_flush_cards = get_unused_flush_cards(&sorted_flush_suit_cards, flush_suit);
        println!("unused_flush_cards {:?} board {:?}", unused_flush_cards, sorted_flush_suit_cards[..3].to_vec());
        vec![(sorted_flush_suit_cards[..3].to_vec(), generate_two_card_combos(&unused_flush_cards))]
    } else {
        Vec::new() // No flush suit found
    }
}

fn get_unused_flush_cards(comcards: &Vec<Card>, suit: Suit) -> Vec<Card> {
    let used_values: Vec<u8> = comcards.iter().map(|c| c.value).collect();
    let all_values: [u8; 13] = [1, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2];
    all_values.iter().filter(
        |v| used_values.contains(v) == false
    ).map(|v| Card {
        value: *v,
        suit
    }).collect::<Vec<Card>>()
} 
pub fn generate_two_card_combos(cards: &[Card]) -> Vec<SameRankHands> {
    let mut combinations: Vec<SameRankHands> = Vec::new();
    for i in 0..cards.len() {
        for j in i + 1..cards.len() {
            combinations.push(vec![vec![cards[i], cards[j]]]);
        }
    }
    combinations
}