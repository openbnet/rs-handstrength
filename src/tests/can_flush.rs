
#[cfg(test)]

mod tests {
    use crate::card::{Card, Suit};
    use crate::can_flush::can_have_flush;
    use insta::assert_debug_snapshot;
    // #[test]
    fn no_flush() {
        let comcards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 3, suit: Suit::D },
            Card { value: 4, suit: Suit::S },
        ];
        let flushes = can_have_flush(&comcards);
        assert_eq!(flushes.len(), 0);
        
    }

    #[test]
    fn flush_exists() {
        let comcards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 3, suit: Suit::H },
            Card { value: 4, suit: Suit::H },
        ];
        let flushes = can_have_flush(&comcards);
        println!("flushes {:?}", flushes.len());
        assert_eq!(flushes.len(), 1);
        assert_eq!(flushes[0].1.len(), 45);
        assert_debug_snapshot!(flushes);
    }
    #[test]
    fn flush_exists4() {
        let comcards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 3, suit: Suit::H },
            Card { value: 4, suit: Suit::H },
            Card { value: 5, suit: Suit::H }
        ];
        let flushes = can_have_flush(&comcards);
        println!("flushes {:?}", flushes.len());
        assert_eq!(flushes.len(), 1);
        assert_eq!(flushes[0].1.len(), 36);
        assert_debug_snapshot!(flushes);
    }
    // // #[test]
    // fn multiple_suits_no_flush() {
    //     let comcards = vec![
    //         Card { value: 2, suit: Suit::H },
    //         Card { value: 3, suit: Suit::D },
    //         Card { value: 4, suit: Suit::H },
    //         Card { value: 5, suit: Suit::C },
    //     ];
    //     assert_eq!(can_have_flush(&comcards), None);
    // }

    // // #[test]
    // fn empty_array() {
    //     let comcards: Vec<Card> = vec![];
    //     assert_eq!(can_have_flush(&comcards), None);
    // }

    // // #[test]
    // fn all_same_suit() {
    //     let comcards = vec![
    //         Card { value: 2, suit: Suit::C },
    //         Card { value: 3, suit: Suit::C },
    //         Card { value: 4, suit: Suit::C },
    //         Card { value: 5, suit: Suit::C },
    //     ];
    //     assert_eq!(can_have_flush(&comcards), Some(Suit::C));
    // }
    
}
