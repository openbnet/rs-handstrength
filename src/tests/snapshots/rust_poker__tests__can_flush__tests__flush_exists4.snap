---
source: src/tests/can_flush.rs
expression: flushes
---
[
    (
        [
            Card {
                value: 5,
                suit: H,
            },
            Card {
                value: 4,
                suit: H,
            },
            Card {
                value: 3,
                suit: H,
            },
        ],
        [
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 13,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 12,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 11,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 10,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 9,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 1,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 12,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 11,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 10,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 9,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 13,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 11,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 10,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 9,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 12,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 11,
                    suit: H,
                },
                Card {
                    value: 10,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 11,
                    suit: H,
                },
                Card {
                    value: 9,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 11,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 11,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 11,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 10,
                    suit: H,
                },
                Card {
                    value: 9,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 10,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 10,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 10,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 9,
                    suit: H,
                },
                Card {
                    value: 8,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 9,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 9,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 8,
                    suit: H,
                },
                Card {
                    value: 7,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 8,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
            [
                Card {
                    value: 7,
                    suit: H,
                },
                Card {
                    value: 6,
                    suit: H,
                },
            ],
        ],
    ),
]
